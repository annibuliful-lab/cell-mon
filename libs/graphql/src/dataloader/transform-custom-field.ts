import { Dictionary, groupBy, orderBy } from 'lodash';

type MapDataWithIdsByCustomFieldId<T> = {
  data: T[];
  ids: string[];
  idField: keyof T;
};

export function mapDataListWithIdsByCustomFieldIds<T>({
  data,
  idField,
  ids,
}: MapDataWithIdsByCustomFieldId<T>) {
  if (data.length === 0) return Array(ids.length).fill(null);

  const dataMap = new Map(
    data.map((d) => [d[idField] as unknown as string, d]),
  );

  return ids.map((id) => dataMap.get(id.toString()));
}

type MapListWithIdsByCustomFieldId<T extends { id: string }> = {
  data: T[];
  ids: string[];
  idField: keyof T;
  keySortOrder?: Dictionary<{
    id: string;
    keyIds: string[];
  }>;
  sortOrderType?: 'asc' | 'desc';
};

export function mapDataListWithIdsByCustomField<T extends { id: string }>({
  data,
  ids,
  idField,
  keySortOrder,
  sortOrderType = 'asc',
}: MapListWithIdsByCustomFieldId<T>) {
  const groupedData = groupBy(data, (tag) => tag[idField]);

  return ids.map((id) => {
    if (!keySortOrder) return groupedData[id];

    const itemSortOrder = keySortOrder[id].keyIds;

    return orderBy(
      groupedData[id],
      (d) => itemSortOrder.findIndex((id) => id === d.id),
      sortOrderType,
    );
  });
}
