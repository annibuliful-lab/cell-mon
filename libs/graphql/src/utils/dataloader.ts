import { Dictionary, groupBy, orderBy } from 'lodash';

export function mapDataloaderWithIds<T extends { id: string }>(
  data: T[],
  ids: string[],
) {
  const dataMap = new Map(data.map((d) => [d.id, d]));
  return ids.map((id) => dataMap.get(id));
}

type DataloaderWithIdsByCustomFieldId<T> = {
  data: T[];
  ids: string[];
  idField: keyof T;
};

export function mapDataloaderWithIdField<T>({
  data,
  idField,
  ids,
}: DataloaderWithIdsByCustomFieldId<T>) {
  if (data.length === 0) return Array(ids.length).fill(null);

  const dataMap = new Map(
    data.map((d) => [d[idField] as unknown as string, d]),
  );

  return ids.map((id) => dataMap.get(id));
}

type MapDatasWithIdsByCustomFieldId<T extends { id: string }> = {
  data: T[];
  ids: string[];
  idField: keyof T;
  keySortOrder?: Dictionary<{
    id: string;
    keyIds: string[];
  }>;
  sorderOrderType?: 'asc' | 'desc';
};

export function mapDataloaderRecords<T extends { id: string }>({
  data,
  ids,
  idField,
  keySortOrder,
  sorderOrderType = 'asc',
}: MapDatasWithIdsByCustomFieldId<T>) {
  const groupedData = groupBy(data, (tag) => tag[idField]);

  return ids.map((id) => {
    if (!keySortOrder) return groupedData[id];

    const itemSortOrder = keySortOrder[id].keyIds;

    return orderBy(
      groupedData[id],
      (d) => itemSortOrder.findIndex((id) => id === d.id),
      sorderOrderType,
    );
  });
}
