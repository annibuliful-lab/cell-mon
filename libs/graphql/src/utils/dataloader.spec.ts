import { mapDataloaderRecord, mapDataloaderRecords } from './dataloader';

describe('dataloader', () => {
  it('returns map data structure with custom field', () => {
    expect(
      mapDataloaderRecord({
        data: [
          { id: '1', name: 'AAAA' },
          { id: '2', name: 'BBBB' },
          { id: 'a', name: 'CCCC' },
        ],
        ids: ['1', '2'],
        idField: 'id',
      }),
    ).toMatchSnapshot();
  });

  it('returns map muliple data structure with custom field', () => {
    expect(
      mapDataloaderRecords({
        data: [
          {
            id: '2860e6ff-87ef-44c8-995f-7c4b516c1794',
            projectId: 'TEST_PROJECT_ID_1',
            tag: 'TAG_A',
          },
          {
            id: '6f5b77dd-7859-4229-86d8-1b92063b24e1',
            projectId: 'TEST_PROJECT_ID_2',
            tag: 'TAG_B',
          },
          {
            id: '2860e6ff-87ef-44c8-995f-7c4b516c1890',
            projectId: 'TEST_PROJECT_ID_1',
            tag: 'TAG_C',
          },
        ],
        ids: ['TEST_PROJECT_ID_1', 'TEST_PROJECT_ID_2'],
        idField: 'projectId',
      }),
    ).toMatchSnapshot();
  });

  it('returns map multiple data structure with sort order', () => {
    expect(
      mapDataloaderRecords({
        data: [
          {
            id: 'tag-1',
            projectId: 'TEST_PROJECT_ID_1',
            tag: 'TAG_A',
          },
          {
            id: 'tag-2',
            projectId: 'TEST_PROJECT_ID_2',
            tag: 'TAG_E',
          },
          {
            id: 'tag-3',
            projectId: 'TEST_PROJECT_ID_1',
            tag: 'TAG_C',
          },
          {
            id: 'tag-4',
            projectId: 'TEST_PROJECT_ID_2',
            tag: 'TAG_B',
          },
          {
            id: 'tag-5',
            projectId: 'TEST_PROJECT_ID_2',
            tag: 'TAG_D',
          },
        ],
        ids: ['TEST_PROJECT_ID_1', 'TEST_PROJECT_ID_2'],
        idField: 'projectId',
        keySortOrder: {
          TEST_PROJECT_ID_1: {
            id: 'TEST_PROJECT_ID_1',
            keyIds: ['tag-3', 'tag-1'],
          },
          TEST_PROJECT_ID_2: {
            id: 'TEST_PROJECT_ID_1',
            keyIds: ['tag-4', 'tag-5', 'tag-2'],
          },
        },
      }),
    ).toMatchSnapshot();
  });
});
