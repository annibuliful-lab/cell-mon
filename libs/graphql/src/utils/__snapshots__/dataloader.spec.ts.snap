// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dataloader mapDataloaderWithIdField returns map data structure with custom field 1`] = `
[
  {
    "id": "1",
    "name": "AAAA",
  },
  {
    "id": "2",
    "name": "BBBB",
  },
]
`;

exports[`dataloader mapDataloaderWithIds returns map data structure 1`] = `
[
  {
    "id": "1",
  },
  {
    "id": "2",
  },
]
`;

exports[`dataloader mapDatasWithIdsByCustomField returns map muliple data structure with custom field 1`] = `
[
  [
    {
      "id": "2860e6ff-87ef-44c8-995f-7c4b516c1794",
      "projectId": "TEST_PROJECT_ID_1",
      "tag": "TAG_A",
    },
    {
      "id": "2860e6ff-87ef-44c8-995f-7c4b516c1890",
      "projectId": "TEST_PROJECT_ID_1",
      "tag": "TAG_C",
    },
  ],
  [
    {
      "id": "6f5b77dd-7859-4229-86d8-1b92063b24e1",
      "projectId": "TEST_PROJECT_ID_2",
      "tag": "TAG_B",
    },
  ],
]
`;

exports[`dataloader mapDatasWithIdsByCustomField returns map multiple data structure with sort order 1`] = `
[
  [
    {
      "id": "tag-3",
      "projectId": "TEST_PROJECT_ID_1",
      "tag": "TAG_C",
    },
    {
      "id": "tag-1",
      "projectId": "TEST_PROJECT_ID_1",
      "tag": "TAG_A",
    },
  ],
  [
    {
      "id": "tag-4",
      "projectId": "TEST_PROJECT_ID_2",
      "tag": "TAG_B",
    },
    {
      "id": "tag-5",
      "projectId": "TEST_PROJECT_ID_2",
      "tag": "TAG_D",
    },
    {
      "id": "tag-2",
      "projectId": "TEST_PROJECT_ID_2",
      "tag": "TAG_E",
    },
  ],
]
`;
